feat: add the openvpn experiment

ainghazal opened the pull request:
# Description

First iteration of a new openvpn experiment. This takes a set of endpoints and uses minivpn to attempt a handshake with each of the configured endpoints.

A new API endpoint has been added to the backend that is able to distribute valid configuration parameters for an OpenVPN connection, including credentials.

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/ooni/probe-cli/blob/master/CONTRIBUTING.md)
- [x] reference issue for this pull request: https://github.com/ooni/probe/issues/2688
- [x] if you changed anything related to how experiments work and you need to reflect these changes in the ooni/spec repository, please link to the related ooni/spec pull request: https://github.com/ooni/spec/pull/293
- [x] if you changed code inside an experiment, make sure you bump its version number

ainghazal commented:
tests green again after latest additions

bassosimone commented on internal/probeservices/openvpn_test.go:
```diff
@@ -0,0 +1,109 @@
+			Type:    "https",
+		},
+	)
+	if err != nil {
+		panic(err) // so fail the test
+	}
```
Please, use `client := runtimex.Try1(NewClient( /* ... */ ))` instead

bassosimone commented on internal/probeservices/openvpn_test.go:
```diff
@@ -0,0 +1,109 @@
+		srv := testingx.MustNewHTTPServer(state.NewMux())
+		defer srv.Close()
+
+		// TODO(ain)
```
```suggestion
		// TODO(ainghazal): PLEASE, EXPLAIN WHAT YOU NEED TO DO HERE
```

Be consistent with the name after the TODO and make sure you explain what is it that needs to be done

bassosimone commented on internal/experiment/openvpn/openvpn_test.go:
```diff
@@ -0,0 +1,421 @@
+			t.Fatalf("expected err=nil, got %v", err)
+		}
+	})
+
```
```suggestion
```

bassosimone commented on internal/experiment/openvpn/openvpn_test.go:
```diff
@@ -0,0 +1,421 @@
+}
+
+func TestAddConnectionTestKeys(t *testing.T) {
+	t.Run("append connection result to empty keys", func(t *testing.T) {
```
Please, add a test for the UDP case as well

ainghazal replied:
added

bassosimone commented on internal/experiment/openvpn/openvpn_test.go:
```diff
@@ -0,0 +1,421 @@
+	if testing.Short() {
+		t.Skip("skip test in short mode")
+	}
+	// TODO -- do a real test, get credentials etc.
```
```suggestion
	// TODO(ainghazal): do a real test, get credentials etc.
```

bassosimone commented on internal/experiment/openvpn/openvpn_test.go:
```diff
@@ -0,0 +1,421 @@
+			t.Fatalf("expected error %v, got %v", someError, err)
+		}
+	})
+
```
```suggestion
```

bassosimone commented on internal/experiment/openvpn/openvpn_test.go:
```diff
@@ -0,0 +1,421 @@
+		Measurement: measurement,
+		Session:     sess,
+	}
+	if sess.Logger() == nil {
+		t.Fatal("logger should not be nil")
+	}
```
Why are we testing for logger to be not-nil here?

ainghazal replied:
no idea, but does not seem relevant to test here.

bassosimone commented on internal/experiment/openvpn/openvpn_test.go:
```diff
@@ -0,0 +1,421 @@
+	if sess.Logger() == nil {
+		t.Fatal("logger should not be nil")
+	}
+	fmt.Println(ctx, args, m)
```
Please, use `t.Log` or `t.Logf` for logging during tests

bassosimone reviewed (changes requested):
My main, still open, design question is about whether we would be able to easily disable the experiment via the OONI backend API if it misbehaves. This is my main reason for saying "request changes". My understanding is that, if the OONI backend does not return any field, then the experiment will use the defaults configured for "riseupvpn". I do not think this is desirable at this stage, because we still think at this experiment as unstable and we would like to have a chance to configure it to not run remotely. How could we achieve this? Would removing the hardcoded fallback be a possibility? 

Apart from that, I have done a careful review that includes:

1. several minimal cosmetic requests including documenting public fields

2. suggestions and questions about simplifying the codebase a bit

3. small design questions

ainghazal commented on internal/engine/inputloader.go:
```diff
@@ -317,6 +335,43 @@ func (il *InputLoader) loadRemote(ctx context.Context) ([]model.OOAPIURLInfo, er
+	for _, provider := range openvpn.APIEnabledProviders {
+		reply, err := il.vpnConfig(ctx, provider)
+		if err != nil {
+			break
```
tx for pointing this out! this for loop is not really used atm, so I'll go  with returning the error

ainghazal commented on internal/engine/inputloader.go:
```diff
@@ -317,6 +335,43 @@ func (il *InputLoader) loadRemote(ctx context.Context) ([]model.OOAPIURLInfo, er
+		if err != nil {
+			break
+		}
+		// here we're just collecting all the inputs. we also cache the configs so that
+		// each experiment run can access the credentials for a given provider.
```
you're right - cacheing happens in the session method. updated the comment.

ainghazal commented on internal/experiment/openvpn/endpoint.go:
```diff
@@ -0,0 +1,278 @@
+	},
+}
+
+// Shuffle returns a shuffled copy of the endpointList.
```
noted. I changed the docstring to reflect that we mutate the list in place.
in any case, I'll probably stop using the default endpoint list, so probably this code needs to be deleted soon.

ainghazal commented on internal/experiment/openvpn/endpoint.go:
```diff
@@ -0,0 +1,278 @@
+	logger model.Logger,
+	endpoint *endpoint,
+	creds *vpnconfig.OpenVPNOptions) (*vpnconfig.Config, error) {
+	// TODO(ainghazal): use merge ability in vpnconfig.OpenVPNOptions merge (pending PR)
```
low prio, but we can look at this after we land the experiments, thanks

ainghazal commented on internal/experiment/openvpn/endpoint_test.go:
```diff
@@ -0,0 +1,402 @@
+		t.Fatal("riseup is the only valid provider now")
+	}
+	if valid := isValidProvider("nsa"); valid {
+		t.Fatal("nsa will nevel be a provider")
```
fixed typo

ainghazal commented on internal/experiment/openvpn/openvpn.go:
```diff
@@ -0,0 +1,384 @@
+)
+
+var (
+	ErrBadAuth = errors.New("bad provider authentication")
```
actually for some reason this is not used anywhere - removed for now.

ainghazal commented on internal/experiment/openvpn/openvpn.go:
```diff
@@ -0,0 +1,384 @@
+		sess.Logger().Warn("Fatal error while attempting to connect to endpoint, aborting!")
+		return err
+	}
+	if connResult != nil {
```
actually on a closer look, I was never returning any error (since any error will be returned as failure). modified the function return types.

ainghazal commented on internal/experiment/openvpn/openvpn.go:
```diff
@@ -0,0 +1,384 @@
+
+// AllConnectionsSuccessful returns true if all the registered handshakes have Status.Success equal to true.
+func (tk *TestKeys) AllConnectionsSuccessful() bool {
+	for _, c := range tk.OpenVPNHandshake {
```
I would say no - added that case.

ainghazal commented on internal/experiment/openvpn/openvpn.go:
```diff
@@ -0,0 +1,384 @@
+	}
+	sess.Logger().Infof("Probing endpoint %s", endpoint.String())
+
+	connResult, err := m.connectAndHandshake(ctx, zeroTime, idx, sess.Logger(), endpoint, openvpnConfig, handshakeTracer)
+	if err != nil {
+		sess.Logger().Warn("Fatal error while attempting to connect to endpoint, aborting!")
+		return err
+	}
```
removed this block, there was no err actually returned after latest code modifications

ainghazal commented on internal/mocks/session.go:
```diff
@@ -70,6 +73,11 @@ func (sess *Session) FetchPsiphonConfig(ctx context.Context) ([]byte, error) {
+func (sess *Session) FetchOpenVPNConfig(
+	ctx context.Context, provider, cc string) (*model.OOAPIVPNProviderConfig, error) {
+	return sess.MockFetchOpenVPNConfig(ctx, provider, cc)
+}
```
added

ainghazal commented on internal/model/archival.go:
```diff
@@ -392,3 +392,38 @@ type ArchivalNetworkEvent struct {
+
+// ArchivalOpenVPNConnectStatus is the status of ArchivalOpenVPNConnectResult.
+type ArchivalOpenVPNConnectStatus struct {
+	Blocked *bool   `json:"blocked,omitempty"`
+	Failure *string `json:"failure"`
+	Success bool    `json:"success"`
```
Dunno. I'm in for flattening, but when loading the results into a tabular form I'd rather have a clear boolean than having to deal with checking for empty strings. But for sure this is something OONI data can do, it depends on the angle. Thinking on someone just analyzing the raw json.

ainghazal commented:
For disabling the experiment: would this work? https://github.com/ooni/probe-cli/pull/1585/commits/8c5fa1181bca51d86dc45b836378c91b64d59be3 

right now if the HTTP client finds a 404 miniooni panics, would this be handled correctly by the mobile/desktop apps just by returning an error?

ainghazal commented:
addressed the comments in the review; I think the only case I'd still defend is adding a boolean for success (in the top-level OpenVPN handshake archival result) since I think simplifies analysis in some cases (ooni data can always override it)

ainghazal commented:
Added a couple of commits, one for accepting options in the oonirun descriptor (not used at the moment), and the other using `measurexlite.NewFailure`. This can be tested with the following oonirun file:

```
{
  "name": "openvpn-riseup",
  "description": "measure vpn connection to riseup gateway",
  "author": "Ain Ghazal <ain@openobservatory.org>",
  "nettests": [
    {
      "test_name": "openvpn",
      "inputs": [
        "openvpn://riseupvpn.corp/?address=51.15.187.53:53&transport=udp"
      ],
      "options": {
        "Cipher": "AES-256-GCM",
        "Auth": "SHA512",
        "Compress": "",
        "Obfuscation": "none",
        "SafeCA": "-----BEGIN CERTIFICATE-----\nMIIBYjCCAQigAwIBAgIBATAKBggqhkjOPQQDAjAXMRUwEwYDVQQDEwxMRUFQIFJv\nb3QgQ0EwHhcNMjExMTAyMTkwNTM3WhcNMjYxMTAyMTkxMDM3WjAXMRUwEwYDVQQD\nEwxMRUFQIFJvb3QgQ0EwWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAQxOXBGu+gf\npjHzVteGTWL6XnFxtEnKMFpKaJkA/VOHmESzoLsZRQxt88GssxaqC01J17idQiqv\nzgNpedmtvFtyo0UwQzAOBgNVHQ8BAf8EBAMCAqQwEgYDVR0TAQH/BAgwBgEB/wIB\nATAdBgNVHQ4EFgQUZdoUlJrCIUNFrpffAq+LQjnwEz4wCgYIKoZIzj0EAwIDSAAw\nRQIgfr3w4tnRG+NdI3LsGPlsRktGK20xHTzsB3orB0yC6cICIQCB+/9y8nmSStfN\nVUMUyk2hNd7/kC8nL222TTD7VZUtsg==\n-----END CERTIFICATE-----",
        "SafeCert": "-----BEGIN CERTIFICATE-----\nMIICdzCCAh2gAwIBAgIQaiLN8Wt8B7n/fkNAb0dPhjAKBggqhkjOPQQDAjAzMTEw\nLwYDVQQDDChMRUFQIFJvb3QgQ0EgKGNsaWVudCBjZXJ0aWZpY2F0ZXMgb25seSEp\nMB4XDTI0MDUyMzExNDYzOFoXDTI0MDYyNzExNDYzOFowFDESMBAGA1UEAxMJVU5M\nSU1JVEVEMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA6/FnGI7yIw0P\nb1p+IRiTrLi+lc2bacboPE//bsTFBh+PuTudyZeGY9mxZDd1wBauwBux8nnSnaKK\nJGzd+5xQrG9xlb19aMSBWpS8+rdCz03zEG31SQgwSAAUjmYsDEhstwY329K7noCZ\n125Dp5HIPZYRS5Xbc5NirazrRongsYdq53Mlmp2S/KPFBo6LkkNppBcSzIo6wsUo\n96PViTy15yA4PWhrI5dP798Wg7od4R5Q/cM+zyaxoycwKUck4Ixhqw7nod8OBXZu\nwITKGGqFmMtAQAa8FZG1y+e85O7GH55K/W/3sO5T4fcV+A9V5AyynRQfZbxvIT0I\nbMX1/y7+WwIDAQABo2cwZTAOBgNVHQ8BAf8EBAMCB4AwEwYDVR0lBAwwCgYIKwYB\nBQUHAwIwHQYDVR0OBBYEFJ81OtiqvnHEnXngCTa2wpyPeOhKMB8GA1UdIwQYMBaA\nFH1KYtj/K0nEebaisdyOPmMHXKj+MAoGCCqGSM49BAMCA0gAMEUCIBmbOvPti2bM\nEkjHRoCRE4h09da1YMDdYhYgndmnQaeMAiEAs0RqX86JTcB2M4FkW+gw1ni3Okuc\ntBKwmlxGufWxzpE=\n-----END CERTIFICATE-----",
        "SafeKey": "-----BEGIN RSA PRIVATE KEY-----\nMIIEogIBAAKCAQEA6/FnGI7yIw0Pb1p+IRiTrLi+lc2bacboPE//bsTFBh+PuTud\nyZeGY9mxZDd1wBauwBux8nnSnaKKJGzd+5xQrG9xlb19aMSBWpS8+rdCz03zEG31\nSQgwSAAUjmYsDEhstwY329K7noCZ125Dp5HIPZYRS5Xbc5NirazrRongsYdq53Ml\nmp2S/KPFBo6LkkNppBcSzIo6wsUo96PViTy15yA4PWhrI5dP798Wg7od4R5Q/cM+\nzyaxoycwKUck4Ixhqw7nod8OBXZuwITKGGqFmMtAQAa8FZG1y+e85O7GH55K/W/3\nsO5T4fcV+A9V5AyynRQfZbxvIT0IbMX1/y7+WwIDAQABAoIBAHLmFUmtWxdcpdaZ\nX/DoEgo70XwMK5Hgbnnoj0C3DCeGOSyAbr+cTbLUcYGXTH1lzmX5Vrf5QWrIm7NP\nXO7J2bOPdeXw6GCbyU5+PmVt11gy4ppuodOV7EUz3M7XzL2Si3a5zXv8bKesgr6Z\nkNLKuJPdP8DqUns/G//txImOXWC13QFk0KGeEzWU5/5tXR8SP/wft68sYQiNmfPw\nxGIuqIs/ILsEYM/vXirBx3Q4KyfrJZ71X3ueRKxhxnMAN6wztYqwa7A9NjSMu566\npjG2c0tBXEk+IrX4ehXk/+tM9oCHM+gV19uOky56mTjNQEetFibMbd8u/5zF/pga\niu+KEBECgYEA/qklejvogI4jFA0pwK//3gYMostn+hq4pM3yBHGgxNuJu2PQr7oP\nZ8KUaHIoZOb/FD2E/JuYkpA6raIWVd/7R+oVwq8OVPpKxtAtjD9u4+gHvkpDF1dv\nNkwHSeG9+9EOcNeqCHzGLK6VJys0FVEBOJSWgGy5xsjvFsYpbrLaJdMCgYEA7S8O\na9RoA0AEr3B/zcrGRKZceyF/z7maOYEUuZL8Cf4GxSciQN5DddcYXl5yzjFUiubV\ncM4kzqvQfBrVYFhNAHOPVQ4Unp+UzEWKe3juPnYVJOh0GLreW0nSC2QMftSZrspC\nlgDr2d3q06tAXDNKo0mko6b4MQF6GzdpNTcpyFkCgYAxutpEuno20IrtGXzz0erH\ncqr5B3uwjZNNK6J9V6srhiupWl6gUlc7zfWpR9G3kpxxWWok4kWzKVMsISD3eBvb\n+UxyjjjgQ1hi5rheUOzYuLD6agob/skK82Hg/aJaEIMfah4cNjGE/DrIQVmUaBMy\n92FEhvboaMi3y86/fVG4XQKBgBBRhX97vLBEjk33woNJKTz96Sz7kAydq3O7Ys6l\nwzt4w8R6vcuSvzdzVhTgEKwJDtUDrrm1JSkm/xAa1IVtbdbTHJBwiJClUBqBylZW\naqXXf/rrF1nAOZ40RQRNnOJ5BB3Xgp9JbvCtaQOpK6NsT/1OCsrLqRXOETWgKVfk\n9LX5AoGAAuuOw6PNpKvef595GzTjsXuGDH7PtVkr4CP5j7qDCKh1t7rjk3RiHzTH\nsyXNZQIsRagE3lo32t5ORhW0X+2oGjAS/YkHKDo/ueGqEzBbWtVcDZKNoxfnjCZ6\ngBNUSqhYykBGDI0ST56CnG0ZJSxztncR39W2TrxpYju3qy1g3Ps=\n-----END RSA PRIVATE KEY-----\n"
      }
    }
  ]
}```

ainghazal commented:
updated the spec to reflect the latest change in the archival format: https://github.com/ooni/spec/pull/293/commits/518c1d5edf1a91a8c0dabbb4847b2e68f07574d1

bassosimone commented on internal/engine/session.go:
```diff
@@ -368,9 +370,35 @@ func (s *Session) FetchTorTargets(
 		return nil, err
 	}
+
+	// here we could also lock the mutex.
```
```suggestion
	// TODO(bassosimone,DecFox): here we could also lock the mutex
	// or we should consider using the same strategy we used for the
	// experiments, where we separated mutable state into dedicated types.
```

bassosimone reviewed (approved):
I am going to merge such that we can continue development in tree. Notably, this experiment is a very good candidate for reasoning about richer input, hence merging it is going to provide collateral help in completing richer input.

💥 🐳
